### PR TITLE
Introduce a bootstrap realm for basic-challenge so it's unique.

### PR DESCRIPTION
**Pull request by sallyom**
/assign @stlaz 

for this: https://github.com/openshift/oc/pull/553

### CODE REVIEW

File: pkg/oauthserver/auth.go [hunk 1]
```diff
@@ -292,7 +292,7 @@ func (c *OAuthServerConfig) getAuthenticationHandler(mux oauthserver.Mux, errorH
 	multiplePasswordProviders := false
 	passwordProviderCount := 0
 	for _, identityProvider := range c.ExtraOAuthConfig.Options.IdentityProviders {
-		if config.IsPasswordAuthenticator(identityProvider) && identityProvider.UseAsLogin {
+		if (config.IsPasswordAuthenticator(identityProvider) || config.IsBootstrap(identityProvider)) && identityProvider.UseAsLogin {
```
Comment (stlaz) [inline pkg/oauthserver/auth.go:295]: no need to change this line

File: pkg/oauthserver/auth.go [hunk 2]
```diff
@@ -308,7 +308,7 @@ func (c *OAuthServerConfig) getAuthenticationHandler(mux oauthserver.Mux, errorH
 		}
 
 		// TODO: refactor handler building per type
-		if config.IsPasswordAuthenticator(identityProvider) {
+		if config.IsPasswordAuthenticator(identityProvider) || config.IsBootstrap(identityProvider) {
```
Comment (stlaz) [inline pkg/oauthserver/auth.go:311]: no need to change this line

File: pkg/oauthserver/auth.go [hunk 3]
```diff
@@ -356,7 +356,12 @@ func (c *OAuthServerConfig) getAuthenticationHandler(mux oauthserver.Mux, errorH
 			}
 			if identityProvider.UseAsChallenger {
 				// For now, all password challenges share a single basic challenger, since they'll all respond to any basic credentials
-				challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger("openshift")
+				// The separate "bootstrap" realm allows 'oc' to distinguish between bootstrap IDP and others
+				if config.IsBootstrap(identityProvider) {
+					challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger("bootstrap")
+				} else {
+					challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger("openshift")
+				}
```
Comment (stlaz) [inline pkg/oauthserver/auth.go:364]: I'd prefer this:
```golang
realm := "openshift"
if config.IsBootstrap(identityProvider) {
    realm = "openshift:kubeadmin"
}
challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger(realm)
```

**Review by stlaz — CHANGES_REQUESTED** [review]

File: pkg/oauthserver/auth.go (hunk 1 above)
Comment (sallyom) [reply pkg/oauthserver/auth.go:295]: fixed, originally i took the bootstrap out of the IsPassword cases :facepalm: thanks

File: pkg/oauthserver/auth.go (hunk 2 above)
Comment (sallyom) [reply pkg/oauthserver/auth.go:311]: fixed, thanks

File: pkg/oauthserver/auth.go (hunk 3 above)
Comment (sallyom) [reply pkg/oauthserver/auth.go]: thanks, updated

Comment (stlaz) [reply pkg/oauthserver/auth.go]: actually, it should probably be
```golang
realm, challengerName := "openshift", "basic-challenge"
if config.IsBootstrap(identityProvider) {
    realm, challengerName = "openshift:kubeadmin", "kubeadmin-challenge"
}
challengers[challengerName] = passwordchallenger.NewBasicAuthChallenger(realm)
```
my bad

Comment (sallyom) [reply pkg/oauthserver/auth.go]: updated

Comment (sallyom) [reply pkg/oauthserver/auth.go]: curious, do we use/process that "basic-challenge"/"kubeadmin-challenge" anywhere? I don't see it.

Comment (stlaz) [reply pkg/oauthserver/auth.go]: the challengers are passed into a union here: https://github.com/openshift/oauth-server/blob/a7f975a5f3dc07f270dc64275033a66308d578a2/pkg/oauthserver/auth.go#L430

What I want is to send different WWW-Authenticate headers for kubeadmin and for normal OpenShift login, but in that case we cannot overwrite the `basic-challenge` challenger with a different one

Comment (sallyom) [reply pkg/oauthserver/auth.go]: i see now, thanks

**Comment by stlaz** [comment]
/hold
this probably was not such a good idea after all - https://github.com/openshift/oc/pull/553#discussion_r489212922

**Comment by sallyom** [comment]
going in a different direction, making the change only in oc, closing this

**Pull request closed without merging**